suite/tests/client-interface/stolen-reg-index.dll.c

```diff
@@ -0,0 +1,142 @@
+        DRREG_SUCCESS) {
+        DR_ASSERT(false);
+    }
+
+    /* XXX i#5498: We should check the address in reg_ptr for correctness. A
+     * constant reference address in this client to check for correctness based
+     * on the test binary stolen-reg-index is not a viable approach because
+     * different versions of compiler, linker/loader and OS will produce
+     * different addresses at runtime.
+     */
+    if (!drutil_insert_get_mem_addr(drcontext, ilist, instr, mref, reg_ptr, reg_tmp)) {
+        dr_printf("drutil_insert_get_mem_addr() failed!\n");
+        return false;
+    }
+
+    if (drreg_unreserve_register(drcontext, ilist, instr, reg_tmp) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+
+    if (drreg_unreserve_register(drcontext, ilist, instr, reg_ptr) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+
+    return true;
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                      bool for_trace, bool translating, void *user_data)
+{
+    opnd_t opnd;
+    reg_id_t reg;
+    reg_id_t stolen = dr_get_stolen_reg();
+    DR_ASSERT(stolen == DR_REG_X28);
+
+    /* Selects store insruction, str x1, [x0, w28, uxtw #0] from test subject,
+     * stolen-reg-index.c.
+     */
+    if (instr_writes_memory(inst)) {
+        opnd = instr_get_dst(inst, 0);
+        if (opnd_is_memory_reference(opnd) && opnd_is_base_disp(opnd)) {
+            if (opnd_get_index(opnd) == stolen && opnd_get_base(opnd) == DR_REG_X0)
+                dr_printf("store memref with index reg X28\n");
+            if (opnd_get_index(opnd) == DR_REG_W28 && opnd_get_base(opnd) == DR_REG_X0)
+                dr_printf("store memref with index reg W28\n");
+            if (insert_get_addr(drcontext, bb, inst, instr_get_dst(inst, 0)))
+                return DR_EMIT_DEFAULT;
+            else
+                DR_ASSERT(false);
+        }
+    }
+
+    /* Selects load instruction, ldr x1, [x0, x28, lsl #0] from test subject,
+     * stolen-reg-index.c.
+     */
+    if (instr_reads_memory(inst)) {
+        opnd = instr_get_src(inst, 0);
+        if (opnd_is_memory_reference(opnd) && opnd_is_base_disp(opnd)) {
+            if (opnd_get_index(opnd) == stolen && opnd_get_base(opnd) == DR_REG_X0)
+                dr_printf("load memref with index reg X28\n");
+            if (opnd_get_index(opnd) == DR_REG_W28 && opnd_get_base(opnd) == DR_REG_X0)
+                dr_printf("load memref with index reg W28\n");
+            if (insert_get_addr(drcontext, bb, inst, instr_get_src(inst, 0)))
+                return DR_EMIT_DEFAULT;
+            else
+                DR_ASSERT(false);
+        }
+    }
+
+    return DR_EMIT_DEFAULT;
+}
+
+DR_EXPORT void
+dr_client_main(client_id_t id, int argc, const char *argv[])
+{
+    drreg_options_t ops = { sizeof(ops), 4 /*max slots needed*/, false };
+    if (!drmgr_init() || !drutil_init() || drreg_init(&ops) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+
```

Comment:
Assert that x28 is the stolen register in case we change the default.
Other tests do this today:
```
suite/tests/client-interface/drreg-test.dll.c:        CHECK(dr_get_stolen_reg() == TEST_REG_STOLEN, "stolen reg doesn't match");
```